types/valuerenderer/valuerenderer.go

```diff
@@ -0,0 +1,207 @@
+package valuerenderer
+
+import (
+	"context"
+	"errors"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+	"fmt"
+
+	"github.com/dustin/go-humanize"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+var RegExp = regexp.MustCompile(`(\d+)(\w+)`)
+
+// ValueRenderer defines an interface to produce formated output for Int,Dec,Coin types as well as parse a string to Coin or Uint.
+type ValueRenderer interface {
+	Format(context.Context, interface{}) (string, error)
+	Parse(context.Context, string) (interface{}, error)
+}
+
+// denomQuerierFunc should take context and baseDenom to invoke DenomMetaData method on queryClient. This method will return banktypes.Metadata.
+type denomQuerierFunc func(context.Context, string) (banktypes.Metadata, error)
+
+// DefaultValueRenderer defines a struct that implements ValueRenderer interface
+type DefaultValueRenderer struct {
+	denomQuerier denomQuerierFunc
+}
+
+var _ ValueRenderer = &DefaultValueRenderer{}
+
+// NewDefaultValueRenderer  initiates denomToMetadataMap field and returns DefaultValueRenderer struct
+func NewDefaultValueRenderer(denomQuerier denomQuerierFunc) DefaultValueRenderer {
+	return DefaultValueRenderer{
+		denomQuerier: denomQuerier,
+	}
+}
+
+// Format converts an empty interface into a string depending on interface type.
+func (dvr DefaultValueRenderer) Format(c context.Context, x interface{}) (string, error) {
+	var sb strings.Builder
+
+	switch v := x.(type) {
+	case types.Dec:
+		s := v.String()
+		if len(s) == 0 {
+			return "", errors.New("empty string")
+		}
+
+		i := strings.Index(s, ".")
+		first, second := s[:i], s[i+1:]
+		first64, err := strconv.ParseInt(first, 10, 64)
+		if err != nil {
+			return "", err
+		}
+
+		sb.WriteString(humanize.Comma(first64))
+		sb.WriteString(".")
+		sb.WriteString(removeTrailingZeroes(second))
+
+	case types.Int:
+		s := v.String()
+		if len(s) == 0 {
+			return "", errors.New("empty string")
+		}
+
+		sb.WriteString(humanize.Comma(v.Int64()))
+
+	case types.Coin:
+		baseDenom, err := baseFromDenom(v.Denom)
+		if err != nil {
+			return "", err
+		}
+
+		metadata, err := dvr.denomQuerier(c, baseDenom)
+		if err != nil {
+			return "", err
+		}
+
+		expSub := computeExponentSubtraction(v.Denom, metadata)
+
+		formatedAmount := dvr.ComputeAmount(v.Amount.Int64(), expSub)
+
+		sb.WriteString(formatedAmount)
+		sb.WriteString(metadata.Display)
+
+	default:
+		panic("type is invalid")
+	}
+
+	return sb.String(), nil
+}
+
+// TODO address the cass where denom starts with "u"
+// baseFromDenom converts denom to banktypes.Metadata.Base that is used in banktypes.QueryDenomMetadataRequest. queryClient.DenomMetadata method takes Banktypes.QueryDenomMetadataRequest and context as arguments and returns resp.Metadata.
+func baseFromDenom(denom string) (string, error) {
```

Comment:
`baseFromDenom` converts `denom` to `banktypes.Metadata.Base` that is used in `banktypes.QueryDenomMetadataRequest` [here](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/value-renderer/types/valuerenderer/valuerenderer_test.go#L150). `queryClient.DenomMetadata` [method](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/value-renderer/types/valuerenderer/valuerenderer_test.go#L154) takes `banktypes.QueryDenomMetadataRequest` and `context` as arguments and returns `resp.Metadata`.